drivers/timer/npcx_itim_timer.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) 2021 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_itim_timer
+
+/**
+ * @file
+ * @brief Nuvoton NPCX kernel device driver for  "system clock driver" interface
+ *
+ * This file contains a kernel device driver implemented by the internal
+ * 64/32-bit timers in Nuvoton NPCX series. Via these two kinds of timers, the
+ * driver provides an standard "system clock driver" interface.
+ *
+ * It includes:
+ * - A system timer based on an ITIM64 (Internal 64-bit timer) instance, clocked
+ *   by APB2 which freq is the same as CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC.
+ * - Provide a 64-bit cycles reading and ticks computation based on it.
+ * - Its prescaler is set to 1 and provide the kernel cycles reading without
+ *   handling overflow mechanism.
+ * - After ec entered "sleep/deep sleep" power state which is used for better
+ *   power consumption, then its clock will stop.
+ *
+ * - A event timer based on an ITIM32 (Internal 32-bit timer) instance, clocked
+ *   by LFCLK which frequency is 32KHz and still activated when ec entered
+ *   "sleep/deep sleep" power state.
+ * - Setup time-out event and handle z_clock_announce() in its ISR.
+ * - Its prescaler is set to 1 and provide timeout event mechanism.
+ * - Compensate system timer which clock is gating after ec left
+ *   "sleep/deep sleep" power state.
+ */
+
+#include <drivers/clock_control.h>
+#include <drivers/timer/system_timer.h>
+#include <sys_clock.h>
+#include <spinlock.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(itim, LOG_LEVEL_ERR);
+
+#define NPCX_ITIM32_MAX_CNT 0xFFFFFFFF
+#define NPCX_ITIM64_MAX_HALF_CNT 0xFFFFFFFF
+#define EVT_CYCLES_PER_SEC LFCLK /* 32768 Hz */
+#define SYS_CYCLES_PER_TICK (sys_clock_hw_cycles_per_sec() \
+					/ CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define EVT_CYCLES_FROM_TICKS(ticks) \
+	ceiling_fraction(ticks * EVT_CYCLES_PER_SEC, \
+			 CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+/* Instance of system and event timers */
+static struct itim64_reg *const sys_tmr = (struct itim64_reg *)
+					DT_INST_REG_ADDR_BY_NAME(0, sys_itim);
+static struct itim32_reg *const evt_tmr = (struct itim32_reg *)
+					DT_INST_REG_ADDR_BY_NAME(0, evt_itim);
+
+static const struct npcx_clk_cfg itim_clk_cfg[] = NPCX_DT_CLK_CFG_ITEMS_LIST(0);
+
+static struct k_spinlock lock;
+/* Announced cycles in system timer before executing z_clock_announce() */
+static uint64_t cyc_sys_announced;
+/* Current target cycles of time-out signal in event timer */
+static uint32_t cyc_evt_timeout;
+
+/* ITIM local inline functions */
+static inline uint64_t npcx_itim_get_sys_cyc64(void)
+{
+	uint32_t cnt64h1, cnt64h2, cnt64l;
```

Comment:
minor point, but for naming I would suggest

cnt64h

and cnt64h_check

then use cnt64h below the loop. That makes it clearer what you are doing

Reply:
Indeed, that's clearer. Thanks.

Reply:
Done.